File: problems/0098/solution.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        res = []

        def dfsInOrder(node):
            if not node:
                return
            dfsInOrder(node.left)
            res.append(node.val)
            dfsInOrder(node.right)

        dfsInOrder(root)
        for i in range(len(res) - 1):
            if res[i] >= res[i + 1]:
                return False
        return True

File: problems/0098/solution.cpp
#include <iostream>
using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

class Solution {
 private:
  void dfsInOrder(TreeNode* root, vector<int>& res) {
    if (root == nullptr) return;
    dfsInOrder(root->left, res);
    res.push_back(root->val);
    dfsInOrder(root->right, res);
  }

 public:
  bool isValidBST(TreeNode* root) {
    if (root == nullptr) return true;
    vector<int> res;
    dfsInOrder(root, res);
    for (int i = 0; i < res.size() - 1; i++) {
      if (res[i] >= res[i + 1]) return false;
    }
    return true;
  }
};